private/pkg/transport/http/httpserver/httpserver.go

```diff
@@ -0,0 +1,150 @@
+	DefaultIdleTimeout = 3 * time.Minute
+)
+
+// Mapper initializes a Router.
+type Mapper interface {
+	Map(chi.Router) error
+}
+
+// MapperFunc is a function for a Mapper
+type MapperFunc func(chi.Router) error
+
+// Map implements Mapper.
+func (m MapperFunc) Map(router chi.Router) error {
+	return m(router)
+}
+
+// NewHTTPHandlerMapper returns a new Mapper for the http.Handler.
+func NewHTTPHandlerMapper(
+	handler http.Handler,
+	options ...HTTPHandlerMapperOption,
+) Mapper {
+	return newHTTPHandlerMapper(handler, options...)
+}
+
+// HTTPHandlerMapperOption is an option for a new HTTPHandlerMapper.
+type HTTPHandlerMapperOption func(*httpHandlerMapper)
+
+// HTTPHandlerMapperWithPrefix returns a new HTTPHandlerMapperOption that uses the prefix.
+func HTTPHandlerMapperWithPrefix(prefix string) HTTPHandlerMapperOption {
+	return func(httpHandlerMapper *httpHandlerMapper) {
+		httpHandlerMapper.prefix = prefix
+	}
+}
+
+// Runner is a runner.
+type Runner interface {
+	// Run runs the router.
+	//
+	// Blocking.
+	// The runner is cancelled when the input context is cancelled.
+	// The listener is closed upon return.
+	//
+	// Response write errors are logged. Response write errors can be ignored.
+	//
+	// Can be called multiple times, resulting in different runs.
+	Run(ctx context.Context, listener net.Listener, mappers ...Mapper) error
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(logger *zap.Logger, options ...RunnerOption) Runner {
+	return newRunner(logger, options...)
+}
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runner)
+
+// RunnerWithShutdownTimeout returns a new RunnerOption that uses the given shutdown timeout.
+//
+// The default is to use DefaultShutdownTimeout.
+// If shutdownTimeout is 0, no graceful shutdown will be performed.
+func RunnerWithShutdownTimeout(shutdownTimeout time.Duration) RunnerOption {
+	return func(runner *runner) {
+		runner.shutdownTimeout = shutdownTimeout
+	}
+}
+
+// RunnerWithReadHeaderTimeout returns a new RunnerOption that uses the given read header timeout.
+//
+// The default is to use DefaultReadHeaderTimeout.
+// If readHeaderTimeout is 0, no read header timeout will be used.
+func RunnerWithReadHeaderTimeout(readHeaderTimeout time.Duration) RunnerOption {
+	return func(runner *runner) {
+		runner.readHeaderTimeout = readHeaderTimeout
+	}
+}
+
+// RunnerWithTLSConfig returns a new RunnerOption that uses the given tls.Config.
+//
+// The default is to use no TLS.
+func RunnerWithTLSConfig(tlsConfig *tls.Config) RunnerOption {
+	return func(runner *runner) {
+		runner.tlsConfig = tlsConfig
+	}
+}
+
+// RunnerWithMiddlewares returns a new RunnerOption that use middlewares when the Runner Run.
+func RunnerWithMiddlewares(middlewares ...func(http.Handler) http.Handler) RunnerOption {
+	return func(runner *runner) {
+		runner.middlewares = append(runner.middlewares, middlewares...)
```

Comment:
Should this be an assignment, e.g., `runner.middlewares = middlewares` or are we appending middleware?

Reply:
assignment works for me, I'm just not sure how do we want to handle when this runner option is called multiple time? assignment will replace the existing middlewares. how do you think?